### PR TITLE
feat(insights): Parameterized first day of the week in queries

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

All our weeks start on Sunday, which is annoyingly US-centric. See ZEN-3780, ZEN-3700, ZEN-3660, ZEN-522, ZEN-3849, ZEN-3833

## Changes

Now in Project Settings:

Screenshot 2023-06-23 at 17 42 40

## How did you test this code?

TODO

### CODE REVIEW

**Comment by raquelmsmith** [comment]
Got another request for this https://posthogusers.slack.com/archives/C01GLBKHKQT/p1689255376235759

File: posthog/api/test/__snapshots__/test_insight_funnels.ambr [hunk 1]
```diff
@@ -1,408 +0,0 @@
-# name: ClickhouseTestFunnelTypes.test_funnel_time_to_convert_auto_bins
```
Comment (thmsobrmlr) [inline posthog/api/test/__snapshots__/test_insight_funnels.ambr:1]: Why is this snapshot missing? The test still seems to be decorated with `@snapshot_clickhouse_queries`.

File: posthog/hogql/database/database.py [hunk 2]
```diff
@@ -92,12 +100,15 @@ def add_warehouse_tables(self, **field_definitions: Any):
             setattr(self, f_name, f_def)
 
 
+BUILT_IN_TABLE_NAMES = [attr for attr in dir(Database) if isinstance(getattr(Database, attr), Table)]
```
Comment (thmsobrmlr) [inline posthog/hogql/database/database.py:103]: This change to table names seems unrelated to the week start date and I don't really understand what it's for - maybe separate this out into a separate PR.

**Review by thmsobrmlr — COMMENTED** [review]
Thanks for tackling this rabbit hole, @Twixes. Definitely needed for our European customers.

The PR was quite large with a number of loosely related changes. I've done my best to review this and left a couple of comments inline. For a ✅ , I'd love if you could:
- Go over the comments to check if something is suspicious (I've left a couple of questions where I was unsure).
- Check the two or three issues mentioned in comments as well.
- Move the changes to HogQL table names to a separate PR or help me understand it.
- It might be good to add a test for each of the insights with start of week = Monday.

File: posthog/hogql/database/database.py (hunk 2 above)
Comment (Twixes) [reply posthog/hogql/database/database.py:103]: Yes, definitely. It was also leftover from a previous HogQL-based approach

File: posthog/api/test/test_insight.py [hunk 3]
```diff
@@ -1248,7 +1248,7 @@ def test_insight_trends_breakdown_pagination(self) -> None:
                     "breakdown_type": "event",
                 },
             )
-        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.json())
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/test/test_insight.py:1251]: Just a debugging aid, previously if the response was not a 200 an extra `print()` was necessary to find out _why_ it wasn't a 200. With this, the response body is automatically shown in such case

File: frontend/src/scenes/project/Settings/WeekStartConfig.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { useActions, useValues } from 'kea'
+import { isAuthenticatedTeam, teamLogic } from 'scenes/teamLogic'
+import { LemonSelect } from '@posthog/lemon-ui'
+import { LemonDialog } from 'lib/lemon-ui/LemonDialog'
+
+export function WeekStartConfig(): JSX.Element {
+    const { currentTeam, currentTeamLoading } = useValues(teamLogic)
+    const { updateCurrentTeam } = useActions(teamLogic)
+
+    const isTeamKnown = isAuthenticatedTeam(currentTeam)
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/project/Settings/WeekStartConfig.tsx:10]: Nope, it's really just about satisfying TypeScript. But looks like this isn't actually needed in this case, removed in #17231

File: posthog/test/test_feature_flag.py [hunk 5]
```diff
@@ -2035,7 +2035,7 @@ def test_cohort_cache_no_unnecessary_queries(self):
         )
         Person.objects.create(team=self.team, distinct_ids=["example_id"], properties={"email": "tim@posthog.com"})
 
-        with self.assertNumQueries(6):
+        with self.assertNumQueries(16):
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/test/test_feature_flag.py:2038]: Very good point, this was because the previous HogQL-based approach required all `Filter`'s to have `team`, which for some reason causes the persons-on-events instance setting to be selected a lot. For now just rolled back the `cohort.py` change

File: posthog/api/test/__snapshots__/test_insight_funnels.ambr (hunk 1 above)
Comment (Twixes) [reply posthog/api/test/__snapshots__/test_insight_funnels.ambr]: Looks like it was some quirk of tests, the snapshot is not changed anymore

**Comment by Twixes** [comment]
Thanks for the great points @thmsobrmlr! They should be addressed here now, but take a good look. To make this change more tolerable, I split out the UI change into #17231.
As for tests, I added a test case for this feature within [test_timezones_weekly](https://github.com/PostHog/posthog/pull/16229/commits/eccfcdac5c9ca271d876c9a7761499d6cddb5106). That's not particularly extensive, but it's the gist of it

File: ee/clickhouse/queries/test/__snapshots__/test_event_query.ambr [hunk 6]
```diff
@@ -74,24 +74,18 @@
...
-            FROM person
-            WHERE team_id = 2
-              AND id IN
-                (SELECT id
-                 FROM person
-                 WHERE team_id = 2
-                   AND ((has(['test'], replaceRegexpAll(JSONExtractRaw(properties, 'name'), '^"|"$', '')))) )
-            GROUP BY id
-            HAVING max(is_deleted) = 0
-            AND ((has(['test'], replaceRegexpAll(JSONExtractRaw(argMax(person.properties, version), 'name'), '^"|"$', ''))))))
```
Comment (thmsobrmlr) [inline ee/clickhouse/queries/test/__snapshots__/test_event_query.ambr:94]: Seems correct, but I can't figure out where this change is coming from.

**Review by thmsobrmlr — APPROVED** [review]
Looks good!

File: ee/clickhouse/queries/test/__snapshots__/test_retention.ambr [hunk 7]
```diff
@@ -17,12 +17,12 @@
         GROUP BY target,
                  event_date),
           target_event_query as
-       (SELECT DISTINCT toStartOfWeek(toDateTime(e.timestamp), 0, 'UTC') AS event_date,
+       (SELECT DISTINCT toStartOfWeek(toTimeZone(toDateTime(e.timestamp, 'UTC'), 'UTC')) AS event_date,
```
[the remark this answers is not in the document]
Comment (Twixes) [reply ee/clickhouse/queries/test/__snapshots__/test_retention.ambr:20]: It's clearer when [`@also_test_with_different_timezones` is used ](https://github.com/PostHog/posthog/pull/16229/files#diff-c3a80c29a590658f38d914aa588ca3a3c91379fe2b692728c75b95bec7717af7L5544)

File: ee/clickhouse/queries/test/__snapshots__/test_event_query.ambr (hunk 6 above)
Comment (Twixes) [reply ee/clickhouse/queries/test/__snapshots__/test_event_query.ambr:94]: I don't know either 🤔 Even though it looks good

Comment (Twixes) [reply ee/clickhouse/queries/test/__snapshots__/test_event_query.ambr:94]: Oh, that's because of this diff https://github.com/PostHog/posthog/pull/16229/files#diff-81b09b4d64b61fe9e989ab1e49f40a4a7885d38e41df15eac395715da1dc3371
Also leftover from the previous HogQL-based approach, reverted

**Review by thmsobrmlr — APPROVED** [review]
Looks good now!
